File: Data Structures & Algorithms/matrixBFS/submission-3.py
class Solution:
    def _insertQ(self, grid: List[List[int]], 
                Q: Deque[tuple[int]], r: int, c: int, 
                rMax: int, cMax: int,  
                visited: set) -> None :

        rInBounds = r >= 0 and r < rMax
        cInBounds = c >= 0 and c < cMax

        if not rInBounds or not cInBounds:
            return
        
        if grid[r][c] != 0 or (r, c) in visited:
            return
        
        visited.add((r, c))
        Q.append((r, c))

    def shortestPath(self, grid: List[List[int]]) -> int:
        visited = set()
        Q = deque()
        
        # important constant variables
        rMax = len(grid)
        cMax = len(grid[0])
        iLOC = [(-1, 0), (1, 0), (0, -1), (0, 1)] # All neigbors in 4 direction

        Q.append((0, 0))
        visited.add((0, 0))

        slen = -1
        while Q:
            slen += 1
            qSize = len(Q)
            for _ in range(qSize):
                r, c = Q.popleft()

                if r == rMax-1 and c == cMax-1:
                    return slen
                for i in range(len(iLOC)):
                    self._insertQ(grid, Q, r+iLOC[i][0], c+iLOC[i][1], rMax, cMax, visited)
            
        return -1
                

File: Data Structures & Algorithms/matrixDFS/submission-3.cpp
class Solution {
public:
    int _dfs(vector<vector<int>>& grid, int r, int c, 
             int rMax, int cMax, vector<vector<bool>> &visited)
    {
        bool rInBounds = r >= 0 && r < rMax;
        bool cInBounds = c >= 0 && c < cMax;
        if(!rInBounds || !cInBounds)
            return 0;

        if(visited[r][c] || grid[r][c] != 0)
            return 0;
        
        if(r == rMax-1 && c == cMax - 1)
            return 1;

        visited[r][c] = true;
        // 4 paths
        int count = 0;
        count += _dfs(grid, r-1, c, rMax, cMax, visited);
        count += _dfs(grid, r+1, c, rMax, cMax, visited);
        count += _dfs(grid, r, c-1, rMax, cMax, visited);
        count += _dfs(grid, r, c+1, rMax, cMax, visited);

        visited[r][c] = false;
        return count;
    }

    int countPaths(vector<vector<int>>& grid) {
        
        int rMax = grid.size();
        int cMax = grid[0].size();
        if(grid[0][0] == 1 || grid[rMax-1][cMax-1] == 1)
            return 0;

        vector<vector<bool>> visited(rMax, vector<bool>(cMax, false));
        return _dfs(grid, 0, 0, rMax, cMax, visited);
    }
};
